начальная версия банкомата

freegate78 opened the pull request:
применил мементо в методах загрузить и получить деньги - если не удалось сделать всю операцию
все остальное сырое, и жду фидбека как это все сделать красиво...
пока других паттернов явно не увидел...

freegate78 committed:
начальная версия банкомата
применил мементо в методах загрузить и получить деньги - если не удалось сделать всю операцию
все остальное сырое, и жду фидбека как это все сделать красиво...
пока других паттернов явно не увидел...

petrelevich commented on HW11-ATM/src/main/java/com/akproductions/atm/Atm.java:
```diff
@@ -0,0 +1,99 @@
+        load_money(cells_load);
+    }
+
+    private void install_cells(Cell [] cells_load){
```
install_cells 
в java так не принято методы именовать
installСells - так по конвенции

petrelevich commented on HW11-ATM/src/main/java/com/akproductions/atm/Atm.java:
```diff
@@ -0,0 +1,99 @@
+
+    private void install_cells(Cell [] cells_load){
+        this.cells = new Cell[cells_load.length];
+        for (int i=0;i<this.cells.length;i++){
```
Обращайте внимание на оформление кода, например на пробелы
Idea делает форматирование автоматически по Ctrl-Alt-l

petrelevich commented on HW11-ATM/src/main/java/com/akproductions/atm/Atm.java:
```diff
@@ -0,0 +1,99 @@
+        }
+    }
+
+    public void put_money(Cell[] cells_put) throws Exception {
```
cells_put
переменные так тоже не именуются
cellsPut - так правильно.

petrelevich commented on HW11-ATM/src/main/java/com/akproductions/cell/Cell.java:
```diff
@@ -0,0 +1,23 @@
+package com.akproductions.cell;
+
+public class Cell {
+    private int nominal;
```
номинал имеет смысл опеределить через enum

freegate78 replied:
я решил сделать наследование от Cupure, добавь в этот класс методы взятия купюр и добавления купюр

petrelevich commented on HW11-ATM/src/main/java/com/akproductions/atm/Atm.java:
```diff
@@ -0,0 +1,99 @@
+public class Atm {
+    private Cell cells [];
+    public Atm(Cell[] cells_load){
+        install_cells(cells_load);
```
Имеет смысл объединить эти два метода в один

```
    private void installCells(Cell[] cells_load) {
        this.cells = new Cell[cells_load.length];
        for (int i = 0; i < this.cells.length; i++) {
            this.cells[i] = new Cell(cells_load[i].getNominal(), cells_load[i].getCount());
        }
    }
```
и создать конструктор.
Так и кода меньше и по циклу два раза бегать не надо и надежность выше.

petrelevich commented on HW11-ATM/src/main/java/com/akproductions/atm/Atm.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    public void put_money(Cell[] cells_put) throws Exception {
+        Map<Integer, Integer> preparedCells = new HashMap(0);
```
не забывайте про generics, и не надо задавать явно некорректную первоначальную емкость.
`Map<Integer, Integer> preparedCells = new HashMap<>();`
Это справедливо и для кода ниже: Map<Integer, Integer> preparedCells

petrelevich commented on HW11-ATM/src/main/java/com/akproductions/atm/Atm.java:
```diff
@@ -0,0 +1,99 @@
+
+        long diff_summa = summa;
+
+        do {
```
очень запутанный для понимания код.
Хорошо бы разбить на методы меньшего размера.
не понятно, как учитывается требование "выдавать запрошенную сумму минимальным количеством банкнот",
сложные алгоритмы не нужны, но попытка выдать сначала крупными купюрами быть должна.

freegate78 committed:
для купюр сделал новый класс - чтобы в нем уже через enum указать какие купюры
в Cell проверяю номинал по enum - и если там нет - ошибка. в этом случае при измении класса купюр - Cell не трогаем
все подметоды выделил в отдельные для лучшего чтения, сделал нужные конструкторы.

petrelevich commented on HW11-ATM/src/main/java/com/akproductions/atm/Atm.java:
```diff
@@ -0,0 +1,117 @@
+
+    }
+
+    private Cell[] getCellsFromHash(Map<Integer, Integer> preparedCells) throws Exception {
```
название метода не очень удачное.
Hash - это от слова хэш-функция
по названию трудно догадаться, что внутри.

freegate78 replied:
согласен, после замены мапы на три - функция сорт ушла в конструктор мапы, и сделал методы в процедуре getMoney - sortCellsByNominal и extractMoneyFromAtm

petrelevich commented on HW11-ATM/src/main/java/com/akproductions/atm/Atm.java:
```diff
@@ -0,0 +1,117 @@
+        }
+    }
+
+    public void putMoney(Cell[] cellsPut) throws Exception {
```
в аргументах putMoney я бы ожидал видеть купюры, а не ячейки

freegate78 replied:
логично, переделал функции get  и put money на работу с типом Cupures []
Также, обнаружил связь между купюрами и ячейками - сделал базовый тип купюра (пачка купюр одного номинала), с проверкой номинала по енаму и выдачей ошибки если номинал не известен, и порожденный тип ячейка - где можно изменять число купюр в пачке, это уже фича работы пачки купюр в ячейке банкомата

petrelevich commented on HW11-ATM/src/main/java/com/akproductions/atm/Atm.java:
```diff
@@ -0,0 +1,117 @@
+        Map<Integer, Integer> preparedCells = new HashMap<Integer, Integer>();
+
+        for (int i = 0; i < cellsPut.length; i++) {
+            int castNominal = 0;
```
тут видимо, имелось в виду,
boolean foundNominal,
т.к. далее следует только проверка на 0.

freegate78 replied:
переделал

petrelevich commented on HW11-ATM/src/main/java/com/akproductions/atm/Atm.java:
```diff
@@ -0,0 +1,117 @@
+        return balance;
+    }
+
+    private List<Map.Entry<Integer, Integer>> sort(Map<Integer, Integer> map, String order) {
```
если для хранения ячеек использовать TreeMap, то можно будет руками не сортировать.

freegate78 replied:
да, поменял на TreeMap

petrelevich commented on HW11-ATM/src/main/java/com/akproductions/atm/Atm.java:
```diff
@@ -0,0 +1,117 @@
+        for (Map.Entry<Integer, Integer> cell : map.entrySet()) {
+            toSort.add(cell);
+        }
+        if (order.equalsIgnoreCase("bigFirst")) {
```
"магические значения" стоит константами заменять.

freegate78 replied:
вынес в начало класса большими буквами

petrelevich commented on HW11-ATM/src/main/java/com/akproductions/cell/Cell.java:
```diff
@@ -0,0 +1,41 @@
+import com.akproductions.cupure.Cupure;
+
+public class Cell {
+    private int nominal;
```
тут можно вместо int использовать enum Nominal

freegate78 replied:
в классе купюр enum остался, но он как валидатор номиналов работает - тьуда я не смог добавить значения кроме одного. поэтому он в конструкторе как проврка номинала используется, а также как массив имен купюр и их номиналов

freegate78 committed:
учел ошибки из комментариев в коммите https://github.com/freegate78/java-2020-03/pull/6/commits/e52bd2e66ba4e5c5f34dbc63f91495de2290af82
-переделал мапу на tree
-сделал класс ячейки наследованием от класса купюр, добавив в него функциональность изменения количества купюр в пачке
-сделал внешние методы банкомата для работы с типом купюры а не ячейки

freegate78 committed:
объект купюра сделал как часть объекта ячейка. вот только получается что есть количество и в объекте купюра и в объекте ячейка. это нормально?
передал объекта компаратора как параметр метода сортировки ячеек по номиналу. для сортировки по умолчанию передаем нулл.

freegate78 committed:
передаю мапу из купюры и ее количества во внешних методах atm.объект купюра имеет только номинал и имя. объект ячейка имеет обхект купуюра и количество их
